crates/control/src/search_suggestor.rs

```diff
@@ -0,0 +1,175 @@
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+    ball_position: Input<Option<BallPosition>, "ball_position?">,
+    invalid_ball_positions: Input<Vec<HypotheticalBallPosition>, "invalid_ball_positions">,
+    robot_to_field: Input<Option<Isometry2<f32>>, "robot_to_field?">,
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    heatmap: AdditionalOutput<DMatrix<f32>, "ball_search_heatmap">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub suggested_search_position: MainOutput<Option<Point2<f32>>>,
+}
+
+impl SearchSuggestor {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        let heatmap_dimensions = (
+            _context.field_dimensions.length.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
+            _context.field_dimensions.width.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
+        );
+        Ok(Self {
+            heatmap_dimensions,
+            heatmap: DMatrix::from_element(heatmap_dimensions.0, heatmap_dimensions.1, 0.0),
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.robot_to_field.copied(),
+            context.search_suggestor_configuration.cells_per_meter,
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.iamax_full();
+        let mut suggested_search_position: Option<Point2<f32>> = None;
+        if self.heatmap.get(maximum_heat_heatmap_position).is_some() {
+            if self.heatmap[maximum_heat_heatmap_position]
+                > context.search_suggestor_configuration.minimum_validity
+            {
+                let mut search_suggestion = Point2::new(
+                    maximum_heat_heatmap_position.0 as f32
+                        / context.search_suggestor_configuration.cells_per_meter as f32,
+                    maximum_heat_heatmap_position.1 as f32
+                        / context.search_suggestor_configuration.cells_per_meter as f32,
+                );
+                let length_half = context.field_dimensions.length / 2.0;
+                let width_half = context.field_dimensions.width / 2.0;
+
+                if search_suggestion.x >= length_half {
+                    search_suggestion.x -= length_half;
+                } else {
+                    search_suggestion.x = length_half - search_suggestion.x
+                }
+                if search_suggestion.y >= width_half {
+                    search_suggestion.y -= width_half;
+                } else {
+                    search_suggestion.y = width_half - search_suggestion.y
+                }
+
+                search_suggestion.x +=
```

Comment:
What is this for?

Reply:
This part was also removed as stated above